### ISSUE TITLE
How do I connect to a Fennel REPL via port?

### ISSUE THREAD

**Issue by ejconlon** [Olical/conjure#531]
Hello - love the project. It works great for managing Fennel REPLs by stdio. However, I am trying to use it to talk to an embedded Fennel REPL listening on a port (https://git.sr.ht/~nasser/8fl for the Renoise tracker), and I can't figure out how to configure Conjure to talk to it with a custom filetype. I know there are builtin clients that work over local sockets - is there an easy way to reuse this functionality? (Could not find the answer in the wiki, issues, etc.) Thank you!

**Comment by russtoku** [comment]
When I read the [Quick Start Guide for the Guile Client](https://github.com/Olical/conjure/wiki/Quick-start:-Guile-(socket)), it looked like it could be forced into service. Unfortunately, it only connects to a named pipe on Windows or a Unix domain socket on Unix and MacOS; not a TCP (network) socket. `:he conjure-client-guile-socket` says:

>Network socket support is planned for this client, but not yet available

The `conjure.remote.socket` module used by the Guile Client says:

>;; TODO Add support for host / port sockets.

Someone would need to write a new Fennel client to connect via a TCP socket. The Guile Client could be used as a model to follow. Support for host / port sockets would need to be added to the `conjure.remote.socket` module, too.

**Comment by ejconlon** [comment]
Thanks for looking into it! It seems like there is no quick and easy way then. I can do what I need to with `iron.nvim` in the meantime (sending lines and visual selections to a subprocess). But I look forward to the Lego approach here: https://github.com/Olical/conjure/discussions/500 !

**Comment by Olical** [comment]
Thanks for looking into this @russtoku, much appreciated! And you're right @ejconlon, Conjure is very tightly coupled inside each client right now, that modularisation work (#500) I have in my head needs to happen ASAP. I'm just focussing on my Clojure debugger right now (prioritising because it'll work with MANY editors and doesn't depend on Conjure, so very high leverage for my time) and almost have that in an alpha state.

Once I pause that for a sec I'll do my Aniseed -> nfnl refactor of Conjure and then move the modules around so we can compose our own clients from reusable units. Sorry it's not there yet but I hope iron.nvim does the trick!

Maybe we could make a thin client that maps through iron.nvim or something so you get all of it's plumbing but you get Conjure's mappings and log UI, if that's what you like.

**Comment by russtoku** [comment]
Just tried `iron.nvim` and it works nicely with Python right out of the box.

In https://github.com/Vigemus/iron.nvim/discussions/264, @jolars suggest the use of treesitter as a way to send a Python function definition to an `iron.nvim` REPL. So, the idea of

> thin client that maps through iron.nvim or something so you get all of it's plumbing but you get Conjure's mappings and log UI

sounds doable.

@ejconlon, are you using https://git.sr.ht/~nasser/8fl-renoise rather than the browser-based version you mentioned?

**Comment by ejconlon** [comment]
Yes, I think the repo has been updated since I linked it; your URL points to the correct version.
